pkg/controllers/management/rkeworkerupgrader/node.go

```diff
@@ -0,0 +1,162 @@
+	}
+
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Status.NodePlan.Plan = nodePlan
+	nodeCopy.Status.NodePlan.Version = cluster.Status.NodeVersion
+	nodeCopy.Status.NodePlan.AgentCheckInterval = nodeserver.AgentCheckIntervalDuringUpgrade
+
+	v3.NodeConditionUpgraded.Unknown(nodeCopy)
+	v3.NodeConditionUpgraded.Message(nodeCopy, msg)
+
+	if _, err := uh.nodes.Update(nodeCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (uh *upgradeHandler) updateNodeActive(node *v3.Node) error {
+	nodeCopy := node.DeepCopy()
+	v3.NodeConditionUpgraded.True(nodeCopy)
+	v3.NodeConditionUpgraded.Message(nodeCopy, "")
+
+	// reset the node
+	nodeCopy.Spec.DesiredNodeUnschedulable = "false"
+	nodeCopy.Status.NodePlan.AgentCheckInterval = nodeserver.DefaultAgentCheckInterval
+
+	if _, err := uh.nodes.Update(nodeCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (uh *upgradeHandler) filterNodes(nodes []*v3.Node, expectedVersion int) (map[string]*v3.Node, map[string]*v3.Node, map[string]*v3.Node, map[string]*v3.Node, int, int, int) {
+	done, upgrading, filtered := 0, 0, 0
+	toPrepareMap, toProcessMap, upgradedMap, notReadyMap := map[string]*v3.Node{}, map[string]*v3.Node{}, map[string]*v3.Node{}, map[string]*v3.Node{}
+
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil || node.Status.NodeConfig == nil || !workerOnly(node.Status.NodeConfig.Role) {
+			continue
+		}
+
+		// skip nodes marked for ignore by user
+		if node.Labels != nil && node.Labels[ignoreKey] == ignoreValue {
+			continue
+		}
+
+		// skip provisioning nodes
+		if !v3.NodeConditionProvisioned.IsTrue(node) || !v3.NodeConditionRegistered.IsTrue(node) {
+			continue
+		}
+
+		filtered++
+
+		// check for nodeConditionReady
+		if !nodehelper.IsMachineReady(node) {
+			notReadyMap[node.Name] = node
+			logrus.Debugf("node [%s] is not ready", node.Name)
+			continue
+		}
+
+		if node.Status.AppliedNodeVersion == expectedVersion {
+			if v3.NodeConditionUpgraded.IsTrue(node) && !node.Spec.InternalNodeSpec.Unschedulable {
+				done++
+			} else {
+				// node hasn't un-cordoned, so consider it upgrading in terms of maxUnavailable count
+				upgrading++
+				upgradedMap[node.Name] = node
+			}
+			continue
+		}
+
+		if preparingNode(node) {
+			// draining or cordoning
+			upgrading++
+			continue
+		}
+
+		if preparedNode(node) {
+			// node ready to upgrade
+			upgrading++
+			toProcessMap[node.Name] = node
+			continue
+		}
+
+		toPrepareMap[node.Name] = node
+	}
+
+	return toPrepareMap, toProcessMap, upgradedMap, notReadyMap, filtered, upgrading, done
+}
+
+func preparingNode(node *v3.Node) bool {
+	return node.Spec.DesiredNodeUnschedulable == "drain" || node.Spec.DesiredNodeUnschedulable == "true"
+}
+
+func preparedNode(node *v3.Node) bool {
+	return v3.NodeConditionDrained.IsTrue(node) || node.Spec.InternalNodeSpec.Unschedulable || v3.NodeConditionUpgraded.IsUnknown(node)
+}
+
+func workerOnly(roles []string) bool {
+	worker := false
```

Comment:
Is `workerOnly` and `isNonWorkerOnly` written differently because you don't want to use `if ! slice.ContainString`?

Reply:
`isNonWorkerOnly` was used https://github.com/rancher/rancher/blob/master/pkg/rkenodeconfigserver/nodeserver.go#L117:1 , so I just got it from there 😄 

I've updated to return in one iteration vs `ContainString` and renamed it to `IsNonWorker` since it checks for etcd/control but doesn't check for a worker role